src/peft/tuners/xlora/model.py

```diff
@@ -0,0 +1,383 @@
+        else:
+            conf = config
+        lora_model = LoraModel(model, config.copy(), adapter_name)
+        self.__dict__["xlora_config"] = conf
+        del self.xlora_config.target_modules
+        self.__dict__["lora_model"] = lora_model
+        super().__init__(model, config, adapter_name)
+        del self.__dict__["lora_model"]
+        del self.__dict__["xlora_config"]
+        self.lora_model = lora_model
+        self.xlora_config = conf
+
+    def post_init_lora(
+        self,
+        model: nn.Module,
+        peft_config: XLoraConfig,
+        model_peft: nn.Module,
+    ) -> None:
+        # model_peft: PeftModel
+        self.xlora_config = peft_config
+
+        if hasattr(model.config, "use_cache") and model.config.use_cache:
+            raise ValueError("`use_cache` must be False")
+
+        adapters_items = peft_config.adapters.items()
+        if hasattr(self.xlora_config, "_subfolders"):
+            adapters_items = zip(peft_config.adapters.items(), self.xlora_config._subfolders)
+        else:
+            adapters_items = peft_config.adapters.items()
+
+        if hasattr(self.xlora_config, "_subfolders"):
+            for (adapter_name, model_id), subfolder in adapters_items:
+                self.lora_model.load_adapter(model_id, adapter_name, subfolder=subfolder)
+        else:
+            for adapter_name, model_id in adapters_items:
+                self.lora_model.load_adapter(model_id, adapter_name)
+
+        self.lora_model.set_adapter(list(peft_config.adapters.keys()))
+
+        def hook(module, *args, **kwargs) -> None:
+            args_real = args[0]
+            kwargs_real: dict = args[1]
+            kwargs_real.update(kwargs)
+
+            xlora_classifier: XLoraClassifier = self.internal_xlora_classifier  # type: ignore
+
+            if "_xlora_classifier_inhibitor_flag" in kwargs_real:
+                payload: InhibitorFlagPayload = kwargs_real["_xlora_classifier_inhibitor_flag"]
+
+                del kwargs_real["_xlora_classifier_inhibitor_flag"]
+
+                self.internal_xlora_scalings = torch.full(  # type: ignore
+                    (payload.batch_size, payload.seq_len, xlora_classifier.n_layers, xlora_classifier.n_classes),
+                    payload.override_scaling_pass_value,
+                )
+
+                return
+
+            xlora_scalings = xlora_classifier.forward(
+                *args_real,
+                **kwargs_real,
+            )
+            # Set the scalings
+            self.internal_xlora_scalings = xlora_scalings
+
+        model.register_forward_pre_hook(hook, with_kwargs=True, prepend=True)
+
+        self._maybe_freeze_all_adapters()
+
+        total_swapped, device = convert_layers_to_xlora(
+            model_peft,
+            self,
+            peft_config,
+        )
+
+        n_classes = len(peft_config.adapters)
+        xlora_classifier = XLoraClassifier(model_peft, peft_config, n_classes, total_swapped, device)
```

Comment:
I think this is a great idea, and I think it is absolutely doable! Thank you.

Reply:
Oh, probably this should also add the `base_model_outputs` in the hook so that it doesn't have to be calculated twice, as it can be re-used, right?

Reply:
I don't think that it is necessary to use the base_model_outputs in the hook because they are only fed into the classifier, am I misunderstanding?

Reply:
No, probably I misunderstand here. But let's just ensure that we don't compute identical outputs like `self.target.base_layer(x, *args, **kwargs)` twice.

Reply:
That is a great optimization, I actually implemented it in mistral.rs but didn't think to do it here for some reason. Do you think caching the output of the base_layer per .forward call would work?

Reply:
Hmm, I'm actually not quite sure anymore. Trying to understand the paper, does the X-LoRA classifier make a prediction based on the hidden states generated by the whole model (i.e. the output from the last layer)? If yes, would this mean that we have to run a complete forward pass with the base model first, calculate the X-LoRA scalings, and then run another complete forward pass with base model + LoRA adapters? This sounds to me like the forward pass would become ~2x as expensive. We could cache the base model output from the first `XLoraLayer`, but after that the cache would be invalid, right, because the inputs to the 2nd layer would change once we apply X-LoRA.

If I misunderstand and the classifier runs separately on each layer, that would be a different story.

Reply:
Yes, the classifier does make a prediction based on the whole model, making the forward pass ~2x more expensive. In [mistral.rs](https://github.com/EricLBuehler/mistral.rs) we implement non-granular scalings which caches the scalings. This is the idea I assume you were talking about, but like you observed we cannot compute `self.target.base_layer` ahead of time. The non-granular scalings method just caches the scalings and avoids the scaling pass, so it is similar in spirit to this idea.

Reply:
Haha, not sure if my idea was quite as advanced, but I think anything that helps reducing overhead is welcome, as I think it could deter some users from trying this out otherwise. If you have some good ideas, if they would deviate from the paper implementation-wise, we could make them optional via an argument, so that default -> paper, non default -> fast.